safehttp/header.go

```diff
@@ -0,0 +1,134 @@
+	}
+	h.immutable[name] = true
+}
+
+// Set sets the header with the given name to the given value.
+// The name is first canonicalized using textproto.CanonicalMIMEHeaderKey.
+// If this headers is not immutable, this function removes all other
+// values currently associated with this header before setting the new
+// value. Returns an error when applied on immutable headers.
+func (h Header) Set(name, value string) error {
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	if disallowedHeaders[name] {
+		return headerIsDisallowedError{name: name}
+	}
+	if h.immutable[name] {
+		return headerIsImmutableError{name: name}
+	}
+	h.wrappedHeader.Set(name, value)
+	return nil
+}
+
+// Add adds a new header with the given name and the given value to
+// the collection of headers. The name is first canonicalized using
+// textproto.CanonicalMIMEHeaderKey. Returns an error when applied
+// on immutable headers.
+func (h Header) Add(name, value string) error {
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	if disallowedHeaders[name] {
+		return headerIsDisallowedError{name: name}
+	}
+	if h.immutable[name] {
+		return headerIsImmutableError{name: name}
+	}
+	h.wrappedHeader.Add(name, value)
+	return nil
+}
+
+// Del deletes all headers with the given name. The name is first
+// canonicalized using textproto.CanonicalMIMEHeaderKey. Returns an
+// error when applied on immutable headers.
+func (h Header) Del(name string) error {
+	name = textproto.CanonicalMIMEHeaderKey(name)
+	if disallowedHeaders[name] {
+		return headerIsDisallowedError{name: name}
+	}
+	if h.immutable[name] {
+		return headerIsImmutableError{name: name}
+	}
+	h.wrappedHeader.Del(name)
+	return nil
+}
+
+// Get returns the value of the first header with the given name.
+// The name is first canonicalized using textproto.CanonicalMIMEHeaderKey.
+// If no header exists with the given name then "" is returned.
+func (h Header) Get(name string) string {
+	return h.wrappedHeader.Get(name)
+}
+
+// Values returns all the values of all the headers with the given name.
+// The name is first canonicalized using textproto.CanonicalMIMEHeaderKey.
+// If no header exists with the name `name` then nil is returned.
+func (h Header) Values(name string) []string {
+	return h.wrappedHeader.Values(name)
+}
+
+// SetCookie adds the cookie provided as a Set-Cookie header in the header
+// collection.
+// TODO: Replace http.Cookie with safehttp.Cookie.
+func (h Header) SetCookie(cookie *http.Cookie) {
+	if v := cookie.String(); v != "" {
+		h.wrappedHeader.Add("Set-Cookie", v)
+	}
+}
+
+// TODO: Add Write, WriteSubset and Clone when needed.
+
+type headerIsImmutableError struct {
+	name string
+}
+
+func (err headerIsImmutableError) Error() string {
+	return "The header with name \"" + err.name + "\" is immutable."
+}
+
+type headerIsDisallowedError struct {
+	name string
+}
+
+func (err headerIsDisallowedError) Error() string {
+	return "The header with name \"" + err.name + "\" is disallowed."
```

Comment:
nit: error strings should not be capitalized, maybe the return value can be:```  `header "` + header_name + `" is disallowed` ``` ?

https://github.com/golang/go/wiki/CodeReviewComments#error-strings